File: resources/recipes/api/cursor-pagination-postgresql/cursor_pagination.py
"""cursor_pagination.py

Python async implementation of cursor-based (keyset) pagination with PostgreSQL.
Uses asyncpg and base64url cursor encoding.
"""

from __future__ import annotations

import base64
import json
from dataclasses import dataclass
from typing import Any, Generic, TypeVar

import asyncpg

T = TypeVar("T")


@dataclass
class CursorData:
    created_at: str
    id: str


@dataclass
class PageResult(Generic[T]):
    data: list[T]
    next_cursor: str | None
    prev_cursor: str | None
    has_more: bool


def encode_cursor(data: CursorData) -> str:
    payload = json.dumps({"createdAt": data.created_at, "id": data.id}).encode()
    return base64.urlsafe_b64encode(payload).decode().rstrip("=")


def decode_cursor(cursor: str) -> CursorData:
    padding = "=" * (-len(cursor) % 4)
    payload = base64.urlsafe_b64decode(cursor + padding).decode()
    parsed = json.loads(payload)
    return CursorData(created_at=parsed["createdAt"], id=parsed["id"])


class PostRepository:
    def __init__(self, pool: asyncpg.Pool) -> None:
        self._pool = pool

    async def find_page(
        self,
        limit: int = 20,
        after_cursor: str | None = None,
        before_cursor: str | None = None,
    ) -> PageResult[dict[str, Any]]:
        async with self._pool.acquire() as conn:
            if after_cursor:
                cursor = decode_cursor(after_cursor)
                rows = await conn.fetch(
                    """
                    SELECT * FROM posts
                    WHERE (created_at, id) < ($1, $2)
                    ORDER BY created_at DESC, id DESC
                    LIMIT $3
                    """,
                    cursor.created_at,
                    cursor.id,
                    limit + 1,
                )
            elif before_cursor:
                cursor = decode_cursor(before_cursor)
                rows = await conn.fetch(
                    """
                    SELECT * FROM (
                      SELECT * FROM posts
                      WHERE (created_at, id) > ($1, $2)
                      ORDER BY created_at ASC, id ASC
                      LIMIT $3
                    ) sub
                    ORDER BY created_at DESC, id DESC
                    """,
                    cursor.created_at,
                    cursor.id,
                    limit + 1,
                )
            else:
                rows = await conn.fetch(
                    """
                    SELECT * FROM posts
                    ORDER BY created_at DESC, id DESC
                    LIMIT $1
                    """,
                    limit + 1,
                )

            has_more = len(rows) > limit
            data = [dict(row) for row in (rows[:limit] if has_more else rows)]

            next_cursor = None
            if has_more and data:
                last = data[-1]
                next_cursor = encode_cursor(
                    CursorData(created_at=str(last["created_at"]), id=str(last["id"]))
                )

            prev_cursor = None
            if data:
                first = data[0]
                prev_cursor = encode_cursor(
                    CursorData(created_at=str(first["created_at"]), id=str(first["id"]))
                )

            return PageResult(
                data=data,
                next_cursor=next_cursor,
                prev_cursor=prev_cursor,
                has_more=has_more,
            )

File: resources/recipes/api/cursor-pagination-postgresql/cursor_pagination.ts
// cursor_pagination.ts
// TypeScript implementation of cursor-based (keyset) pagination with PostgreSQL.
// Uses the `pg` driver and base64url cursor encoding.

import { Pool } from 'pg';
import { Buffer } from 'buffer';

interface CursorData {
  createdAt: string;
  id: string;
}

interface PageResult<T> {
  data: T[];
  nextCursor: string | null;
  prevCursor: string | null;
  hasMore: boolean;
}

interface Post {
  id: string;
  created_at: string;
  title: string;
  score: number;
}

function encodeCursor(data: CursorData): string {
  const json = JSON.stringify(data);
  return Buffer.from(json).toString('base64url');
}

function decodeCursor(cursor: string): CursorData {
  const json = Buffer.from(cursor, 'base64url').toString('utf8');
  return JSON.parse(json);
}

class PostRepository {
  constructor(private pool: Pool) {}

  async findPage(
    limit: number = 20,
    afterCursor?: string,
    beforeCursor?: string
  ): Promise<PageResult<Post>> {
    const client = await this.pool.connect();

    try {
      let query: string;
      let params: unknown[];

      if (afterCursor) {
        const { createdAt, id } = decodeCursor(afterCursor);
        query = `
          SELECT * FROM posts
          WHERE (created_at, id) < ($1, $2)
          ORDER BY created_at DESC, id DESC
          LIMIT $3
        `;
        params = [createdAt, id, limit + 1];
      } else if (beforeCursor) {
        const { createdAt, id } = decodeCursor(beforeCursor);
        query = `
          SELECT * FROM (
            SELECT * FROM posts
            WHERE (created_at, id) > ($1, $2)
            ORDER BY created_at ASC, id ASC
            LIMIT $3
          ) sub
          ORDER BY created_at DESC, id DESC
        `;
        params = [createdAt, id, limit + 1];
      } else {
        query = `
          SELECT * FROM posts
          ORDER BY created_at DESC, id DESC
          LIMIT $1
        `;
        params = [limit + 1];
      }

      const result = await client.query(query, params);
      const rows = result.rows;
      const hasMore = rows.length > limit;
      const data = hasMore ? rows.slice(0, limit) : rows;

      const nextCursor =
        hasMore && data.length > 0
          ? encodeCursor({
              createdAt: data[data.length - 1].created_at,
              id: data[data.length - 1].id,
            })
          : null;

      const prevCursor =
        data.length > 0
          ? encodeCursor({ createdAt: data[0].created_at, id: data[0].id })
          : null;

      return {
        data,
        nextCursor,
        prevCursor: afterCursor || (!beforeCursor && data.length > 0) ? prevCursor : null,
        hasMore,
      };
    } finally {
      client.release();
    }
  }
}

export { PostRepository, encodeCursor, decodeCursor };
export type { CursorData, PageResult, Post };
